helix-core/src/syntax.rs

```diff
@@ -1856,137 +1856,347 @@ fn injection_for_match<'a>(
     (language_name, content_node, included_children)
 }
 
-pub struct Merge<I> {
-    iter: I,
-    spans: Box<dyn Iterator<Item = (usize, std::ops::Range<usize>)>>,
+struct Merge<I: Iterator<Item = HighlightEvent>> {
+    left: iter::Peekable<I>,
+    right: iter::Peekable<Box<dyn Iterator<Item = HighlightEvent>>>,
 
-    next_event: Option<HighlightEvent>,
-    next_span: Option<(usize, std::ops::Range<usize>)>,
+    left_queue: Vec<HighlightEvent>,
+    right_queue: Vec<HighlightEvent>,
 
-    queue: Vec<HighlightEvent>,
+    merge_queue: VecDeque<HighlightEvent>,
 }
 
-/// Merge a list of spans into the highlight event stream.
+/// Merges two HighlightEvent iterators.
+///
+/// Highlights from `right` overlay highlights from `left`.
+/// Any `Source` events in `right` which preceed or go beyond any `Source`
+/// events in `left` are truncated or discarded.
+///
+/// If all of the following invariants are satisfied for `left` and `right`
+/// then the invariants hold for the resulting merged iterator as well.
+///
+/// - `Source` events are sorted by `start`, monotonically increasing.
+///   Ignoring `HighlightStart` and `HighlightEnd` events, for all
+///   consecutive `l` and `r`, `l.start < r.start`.
```

Comment:
We could have a couple of `debug_assert!`s so that this at least gets caught on debug builds or tests

Reply:
I introduced two new `Option<std::ops::Range<usize>>` fields on the `Merge` to track these two. If I understand `debug_assert_eq!` correctly, we never end up reading or writing to those fields in release builds so I think the impact should be minimal.

Pretty handy that this works, thanks for the pointer 😀